File: src/main/java/cinema/domain/Cinema.java
/* This code was initialised by Akka Serverless tooling.
 * As long as this file exists it will not be re-generated.
 * You are free to make changes to this file.
 */

package cinema.domain;

import cinema.CinemaApi;
import com.akkaserverless.javasdk.EntityId;
import com.akkaserverless.javasdk.Reply;
import com.akkaserverless.javasdk.eventsourcedentity.*;
import com.google.protobuf.Empty;

/** An event sourced entity. */
@EventSourcedEntity(entityType = "cinema")
public class Cinema extends AbstractCinema {
    @SuppressWarnings("unused")
    private final String entityId;

    public Cinema(@EntityId String entityId) {
        this.entityId = entityId;
    }

    @Override
    public CinemaDomain.CinemaState snapshot() {
        // TODO: produce state snapshot here
        return CinemaDomain.CinemaState.newBuilder().build();
    }

    @Override
    public void handleSnapshot(CinemaDomain.CinemaState snapshot) {
        // TODO: restore state from snapshot here
    }

    @Override
    public Reply<Empty> increase(CinemaApi.CIncreaseValue command, CommandContext ctx) {
        return Reply.failure("The command handler for `Increase` is not implemented, yet");
    }

    @Override
    public Reply<Empty> decrease(CinemaApi.CDecreaseValue command, CommandContext ctx) {
        return Reply.failure("The command handler for `Decrease` is not implemented, yet");
    }

    @Override
    public Reply<Empty> reset(CinemaApi.CResetValue command, CommandContext ctx) {
        return Reply.failure("The command handler for `Reset` is not implemented, yet");
    }

    @Override
    public Reply<CinemaApi.CurrentCinema> getCurrentCinema(CinemaApi.CGetCinema command, CommandContext ctx) {
        return Reply.failure("The command handler for `GetCurrentCinema` is not implemented, yet");
    }


}

File: src/main/scala/sample/domain/inner/Cinema.scala
package sample.domain.inner

import cinema.CinemaApi
import cinema.domain.CinemaDomain.AudienceAdded
import cinema.domain.{AbstractCinema, CinemaDomain}
import com.akkaserverless.javasdk.Reply
import com.akkaserverless.javasdk.eventsourcedentity._
import com.google.protobuf.Empty

@EventSourcedEntity(entityType = "cinema")
class Cinema(private var entityId: String) extends AbstractCinema {
  override def snapshot(): CinemaDomain.CinemaState =
    return CinemaDomain.CinemaState.newBuilder.build

  override def handleSnapshot(snapshot: CinemaDomain.CinemaState): Unit = {}

  override def increase(
      command: CinemaApi.CIncreaseValue,
      ctx: CommandContext
  ): Reply[Empty] = {
    ctx.emit(AudienceAdded.newBuilder().setId(???).build())
    Reply.message(Empty.getDefaultInstance)
  }

  override def decrease(
      command: CinemaApi.CDecreaseValue,
      ctx: CommandContext
  ): Reply[Empty] = ???

  override def reset(
      command: CinemaApi.CResetValue,
      ctx: CommandContext
  ): Reply[Empty] = ???

  override def getCurrentCinema(
      command: CinemaApi.CGetCinema,
      ctx: CommandContext
  ): Reply[CinemaApi.CurrentCinema] = ???
}
